doc/src/extras.rst

```diff
@@ -141,6 +141,383 @@ Logging cursor
+
+        If *reply* is `!True` sends a keepalive message in either case.
+
+        Returns `!True` if the feedback message was sent successfully,
+        `!False` otherwise.
+
+    Low-level methods for asynchronous connection operation.
+
+    With the synchronous connection, a call to `consume_replication_stream()`
+    handles all the complexity of handling the incoming messages and sending
+    keepalive replies, but at times it might be beneficial to use low-level
+    interface for better control, in particular to `~select.select()` on
+    multiple sockets.  The following methods are provided for asynchronous
+    operation:
+
+    .. method:: read_replication_message(decode=True)
+
+        :param decode: a flag indicating that unicode conversion should be
+                       performed on the data received from the server
+
+        This method should be used in a loop with asynchronous connections
+        after calling `start_replication()` once.
+
+        It tries to read the next message from the server, without blocking
+        and returns an instance of `ReplicationMessage` or `!None`, in case
+        there are no more data messages from the server at the moment.
+
+        It is expected that the calling code will call this method repeatedly
+        in order to consume all of the messages that might have been buffered,
+        until `!None` is returned.  After receiving a `!None` value from this
+        method, the caller should use `~select.select()` or `~select.poll()`
+        on the corresponding connection to block the process until there is
+        more data from the server.
+
+        The server can send keepalive messages to the client periodically.
+        Such messages are silently consumed by this method and are never
+        reported to the caller.
+
+    .. method:: fileno()
+
+        Call the corresponding connection's `~connection.fileno()` method and
+        return the result.
+
+        This is a convenience method which allows replication cursor to be
+        used directly in `~select.select()` or `~select.poll()` calls.
+
+    .. attribute:: replication_io_timestamp
+
+        A `~datetime` object representing the timestamp at the moment of last
+        communication with the server (a data or keepalive message in either
+        direction).
+
+    An actual example of asynchronous operation might look like this::
+
+        keepalive_interval = 10.0
+        while True:
+            msg = cur.read_replication_message()
+            if msg:
+                consumer.consume(msg)
+            else:
+                timeout = keepalive_interval - (datetime.now() - cur.replication_io_timestamp).total_seconds()
+                if timeout > 0:
+                    sel = select.select([cur], [], [], timeout)
+                else:
+                    sel = []
+
+                if not sel:
+                    cur.send_replication_feedback()
```

Comment:
On timeout, `select.select()` returns `([], [], [])` which evaluates as `True` since it's a non-empty tuple. I think this should probably be

```
                else:
                    sel = ([],[],[])

                if not sel[0]:
                    cur.send_replication_feedback()
```